central/declarativeconfig/manager_impl.go

```diff
@@ -224,35 +236,54 @@ func (m *managerImpl) runReconciliation() {
+	m.doUpsert(transformedMessagesByHandler)
+	m.doDeletion(transformedMessagesByHandler)
+}
+
+func (m *managerImpl) doUpsert(transformedMessagesByHandler map[string]protoMessagesByType) {
 	for _, protoType := range protoTypesOrder {
 		for handler, protoMessagesByType := range transformedMessagesByHandler {
 			messages, hasMessages := protoMessagesByType[protoType]
 			if !hasMessages {
 				continue
 			}
-			typeUpdater, hasUpdater := m.updaters[protoType]
-			if !hasUpdater {
-				m.handleMissingTypeUpdater(handler, protoType, messages)
-				return
-			}
+			typeUpdater := m.updaters[protoType]
 			for _, message := range messages {
 				err := typeUpdater.Upsert(m.reconciliationCtx, message)
 				m.updateHealthForMessage(handler, message, err, consecutiveReconciliationErrorThreshold)
 			}
 		}
 	}
-	log.Debugf("Deleting all proto messages that have traits.Origin==DECLARATIVE but are not contained"+
-		" within the current list of transformed messages: %+v", transformedMessagesByHandler)
 }
 
-func (m *managerImpl) handleMissingTypeUpdater(handler string, protoType reflect.Type, messages []proto.Message) {
-	err := fmt.Errorf("manager does not have updater for type %v", protoType)
-	for _, message := range messages {
-		// Set the threshold to 0, meaning we will _always_ update the integration health status to unhealthy.
-		m.updateHealthForMessage(handler, message, err, 0)
+func (m *managerImpl) doDeletion(transformedMessagesByHandler map[string]protoMessagesByType) {
+	reversedProtoTypes := sliceutils.Reversed(protoTypesOrder)
+	var allProtoIDsToSkip []string
+	for _, protoType := range reversedProtoTypes {
+		var idsToSkip []string
+		for _, protoMessageByType := range transformedMessagesByHandler {
+			messages := protoMessageByType[protoType]
+			for _, message := range messages {
+				idsToSkip = append(idsToSkip, m.idExtractor(message))
+			}
+		}
+		allProtoIDsToSkip = append(allProtoIDsToSkip, idsToSkip...)
+		typeUpdater := m.updaters[protoType]
+		log.Debugf("Running deletion with resource updater %T, skipping IDs %+v", typeUpdater, idsToSkip)
+		failedDeletionIDs, err := typeUpdater.DeleteResources(m.reconciliationCtx, idsToSkip...)
+		log.Debugf("Finished deletion, return value: %+v", err)
+		// In case of an error, ensure we do not delete the integration health status for resources we failed to delete.
+		// Otherwise, the reason why the deletion failed will not be visible to users while the resource may still
+		// exist.
+		if err != nil {
+			log.Debugf("The following IDs failed deletion: [%s]", strings.Join(failedDeletionIDs, ","))
+			allProtoIDsToSkip = append(allProtoIDsToSkip, failedDeletionIDs...)
+		}
+	}
+
+	if err := m.removeStaleHealthStatus(allProtoIDsToSkip); err != nil {
+		log.Errorf("Failed to delete stale health status entries for declarative config: %v", err)
 	}
-	utils.Should(err)
-	m.stopSignal.Signal()
```

Comment:
Do we have plans to use `stopSignal` within MVP? If not, let's delete it completely

Reply:
Yeah, don't think so I guess. I removed it, seems simple enough to re-add it if we see any value in it.